### PR TITLE
UI: Alloc FS: File Viewer

### PR DESCRIPTION
**Pull request by DingoEatingFuzz**
Completes the feature set of the allocation explorer by making files viewable. File viewing currently supports:

1. Text files (logs, json, code, etc.)
2. Images (jpg, png, svg, webp, gif, etc.)

If a file is not supported by the viewer, there will still be a link to open the API endpoint in a new tab for convenience. Depending on the file, the browser will either embed it (e.g., pdf, video) or download it for you (arbitrary binary).

image
_A text file shown in the file cat view_

image
_Tailing a file in the file viewer_

image
_An embedded image from within an allocation_

image
_An unsupported file type_

### CODE REVIEW

File: ui/tests/integration/image-file-test.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    // HACK!
+    // Without mocking the entire image element, this component will make a
+    // real request for a real image. It's unfortuante to couple this test to the
+    // favicon, but it's an unlikely asset to be removed, and it's better than
+    // attempting to mock this.
+    src: '/ui/favicon.png',
+    alt: 'This is the alt text',
+    size: 123456,
+  };
+
```
Comment (johncowen) [inline ui/tests/integration/image-file-test.js]: Would using a data-uri or a Blob here? Loading an svg using a data-uri would be the simplest I think. I'm not sure what you need to do exactly here so I might be going down the wrong path - but I think a datauri would work for what you need..maybe..?

File: ui/tests/unit/adapters/node-test.js [hunk 2]
```diff
@@ -20,7 +20,6 @@ module('Unit | Adapter | Node', function(hooks) {
     this.server.create('allocation', { id: 'node-1-2', nodeId: 'node-1' });
     this.server.create('allocation', { id: 'node-2-1', nodeId: 'node-2' });
     this.server.create('allocation', { id: 'node-2-2', nodeId: 'node-2' });
-    this.server.logging = true;
```
Comment (backspace) [inline ui/tests/unit/adapters/node-test.js]: I was wondering about this!

**Comment by backspace** [comment]
oh yes, one other thing I meant to include, is that I think the spacing in the breadcrumbs looks a bit off when viewing a file vs a directory

ls-spacing

file-spacing

I’m not sure what’s preferable: make the container the same height for both to avoid jumping when browsing from a directory into a file, or have the text move down when it’s a file…??

**Comment by johncowen** [comment]
Hey folks the other thing I've noticed here (thanks for the grabs @backspace !), there's a lot of testing for emojis going on here, which I believe are URL safe (yet multibyte).

Are there tests for paths/filenames with things like spaces and question marks, I can't see anything here? I also had a quick look in the other PRs related to this but I didn't see anything (but I might have missed it).

Is this something being tested for elsewhere?

John

File: ui/tests/integration/image-file-test.js (hunk 1 above)
Comment (DingoEatingFuzz) [reply ui/tests/integration/image-file-test.js]: Ah cool, thank you both for investigating. I assumed poorly that a data-uri wouldn't trigger the onload event, but in hindsight it makes sense that the API is consistently abstract from the data scheme.

File: ui/app/templates/components/image-file.hbs [hunk 3]
```diff
@@ -0,0 +1,11 @@
+<a data-test-image-link href={{src}} target="_blank" rel="noopener noreferrer" class="image-file-image">
+  <img data-test-image src={{src}} alt={{or alt fileName}} title={{fileName}} onload={{action updateImageMeta}} />
+</a>
+<figcaption class="image-file-caption">
+  <span class="image-file-caption-primary">
+    <strong data-test-file-name>{{fileName}}</strong>
+    {{#if (and width height)}}
+      <span data-test-file-stats>({{width}}px &times; {{height}}{{#if size}}px, {{format-bytes size}}{{/if}})</span>
```
[the remark this answers is not in the document]
Comment (DingoEatingFuzz) [reply ui/app/templates/components/image-file.hbs]: Yep 🙈

File: ui/app/styles/components/image-file.scss [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  .image-file-caption-primary {
+    display: block;
+    color: $grey;
+  }
+
+  .image-file-caption-secondary {
+    display: block;
+    color: $grey-light;
+    font-style: italic;
+  }
```
[the remark this answers is not in the document]
Comment (DingoEatingFuzz) [reply ui/app/styles/components/image-file.scss]: Indeed. Will delete.

File: ui/app/templates/components/task-file.hbs [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+</div>
+<div data-test-file-box class="boxed-section-body {{if (eq fileComponent "stream") "is-dark is-full-bleed"}}">
+  {{#if (eq fileComponent "stream")}}
+    {{streaming-file logger=logger mode=mode isStreaming=isStreaming}}
+  {{else if (eq fileComponent "image")}}
+    {{image-file src=catUrl alt=stat.Name size=stat.Size}}
+  {{else}}
+    <div data-test-unsupported-type class="empty-message is-hollow">
+      <h3 class="empty-message-headline">Unsupported File Type</h3>
+      <p class="empty-message-body message">The Nomad UI could not render this file, but you can still call view the file directly.</p>
```
[the remark this answers is not in the document]
Comment (DingoEatingFuzz) [reply ui/app/templates/components/task-file.hbs]: Um, how it got there is lost to time, but I'll remove it 🔪

File: ui/app/components/streaming-file.js [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+  stream: task(function*() {
+    // Follow the log if the scroll position is near the bottom of the cli window
+    this.logger.on('tick', () => {
+      run.scheduleOnce('afterRender', () => this.synchronizeScrollPosition());
+    });
+
+    // Force the scroll position to the bottom of the window when starting streaming
+    this.logger.one('tick', () => {
+      run.scheduleOnce('afterRender', () => this.synchronizeScrollPosition(true));
+    });
```
[the remark this answers is not in the document]
Comment (DingoEatingFuzz) [reply ui/app/components/streaming-file.js]: I agree, and I appreciate your attention to detail.

File: ui/app/templates/allocations/allocation/task/fs.hbs [hunk 7]
```diff
@@ -1,31 +1,45 @@
...
+                  {{/link-to}}
+                </li>
+              {{/each}}
+            </ul>
+          </nav>
+        {{/task-file}}
+    {{else}}
+      <div class="fs-explorer boxed-section">
+        <div class="boxed-section-head">
+          <nav class="breadcrumb" data-test-fs-breadcrumbs>
```
[the remark this answers is not in the document]
Comment (DingoEatingFuzz) [reply ui/app/templates/allocations/allocation/task/fs.hbs]: Yeah, I thought the same thing, then thought that duplication isn't that bad, then decided I didn't want to touch anything I didn't need to...

...then did way more than I needed to in Mirage.

File: ui/app/templates/components/task-file.hbs [hunk 8]
```diff
@@ -0,0 +1,39 @@
+{{#if noConnection}}
+  <div data-test-connection-error class="notification is-error">
+    <h3 class="title is-4">Cannot fetch file</h3>
+    <p>The files for this task are inaccessible. Check the condition of the client the allocation is on.</p>
+  </div>
+{{/if}}
```
[the remark this answers is not in the document]
Comment (DingoEatingFuzz) [reply ui/app/templates/components/task-file.hbs]: Yeah, this is definitely gross. I'll look into it.

File: ui/app/templates/components/task-file.hbs [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+      <a data-test-log-action="raw" class="button is-white" href="{{catUrl}}" target="_blank" rel="noopener noreferrer">View Raw File</a>
+    {{/if}}
+    {{#if (and isLarge isStreamable)}}
+      <button data-test-log-action="head" class="button is-white" onclick={{action "gotoHead"}}>Head</button>
+      <button data-test-log-action="tail" class="button is-white" onclick={{action "gotoTail"}}>Tail</button>
+    {{/if}}
+    {{#if isStreamable}}
+      <button data-test-log-action="toggle-stream" class="button is-white" onclick={{action "toggleStream"}}>
+        {{x-icon (if logger.isStreaming "media-pause" "media-play") class="is-text"}}
+      </button>
```
[the remark this answers is not in the document]
Comment (DingoEatingFuzz) [reply ui/app/templates/components/task-file.hbs]: Interesting, must be some sort of `width:auto` situation combined with a long log line? I'll look into this too.

File: ui/tests/unit/adapters/node-test.js (hunk 2 above)
Comment (DingoEatingFuzz) [reply ui/tests/unit/adapters/node-test.js]: This deletion was special enough to get its own commit 🤓 

It was already in the code (as evident by the diff) and I just happened to notice it spewing log lines in the test runner. 

If you aren't familiar with `this.server.logging = true`, it logs every HTTP request that goes through Pretender. Great for debugging, super noisy though.

Comment (backspace) [reply ui/tests/unit/adapters/node-test.js]: yeah, what do you think of [this kind of solution](https://github.com/backspace/prison-rideshare-ui/blob/primary/mirage/config.js#L7) that lets you add a query parameter to turn on logging? I like it more than having to edit code 😆

Comment (DingoEatingFuzz) [reply ui/tests/unit/adapters/node-test.js]: ooOOOoo, that's quite neat indeed. I like that. And nothing about the qunit runner removes that query param in an annoying manner?

Comment (backspace) [reply ui/tests/unit/adapters/node-test.js]: I haven’t had any trouble with it so far! 🎉

Comment (johncowen) [reply ui/tests/unit/adapters/node-test.js]: +1 to doing this. We do this in consul lots, but we use `document.cookies` instead. Using cookies means settings stick around when the URL changes, which is handy. Then we use Web Inspector to edit the cookies, which turns the Web Inspector into kind of a semi 'dev preferences' panel.

Actually, we didn't ever add a setting to turn HTTP logging on and off (we use it more to control mocks/doubles), but I think I'd add that also as we have some noisy logging going on aswell which I don't mind but other folks might.

**Comment by DingoEatingFuzz** [comment]
Ooookay, all comments addressed!

Everything is in new commits, so that's probably the easiest way to re-review, but I'll go over some of the bigger things.

- **Errant No Connection error**: This was due to the allocation's client not being loaded in time, which I fixed and added a test for.
- **File cat can go off screen**: This was some CSS that was oddly related to the clipped tooltip `overflow-x:auto` bug from a couple months ago. I set the width while accounting for the margin and it's good now.
- **Breadcrumbs have awkward bottom-margin when viewing a file**: This was due to the button height, which was no secret. I fixed it with an existing style that I forget existed: `is-compact`. This modifier also needed to be tweaked since this is the first time it's being used for multiple buttons in a bar.
- **Missing test coverage for tricky characters like "?"**: I added a new set of possible characters to the mirage factory. Everything is still working fine!

**Comment by johncowen** [comment]
Hey @DingoEatingFuzz ,

I gave the URL chars a try, some work, but I could definitely get things to go awry with `%`'s at least. I could reasonably consistently get a 'URI malformed' to appear.

Screenshot 2019-08-09 at 09 25 48


Don't let this hold you up though, if you want to move on with this and look at that later.

John

File: ui/app/components/task-file.js [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+    const timing = this.useServer ? this.serverTimeout : this.clientTimeout;
+    const logFetch = url =>
+      RSVP.race([this.token.authorizedRequest(url), timeout(timing)]).then(
+        response => {
+          if (!response || !response.ok) {
+            this.nextErrorState(response);
+          }
+          return response;
+        },
+        error => this.nextErrorState(error)
```
Comment (backspace) [inline ui/app/components/task-file.js]: Nice fix! 🤩

File: ui/app/templates/allocations/allocation/task/fs.hbs [hunk 11]
```diff
@@ -1,31 +1,15 @@
...
-        <div class="boxed-section-body">
-          <div data-test-file-viewer>placeholder file viewer</div>
+    {{#if isFile}}
+      {{#task-file allocation=task.allocation task=task file=path stat=stat class="fs-explorer"}}
+        {{fs-breadcrumbs task=task path=path}}
+      {{/task-file}}
+    {{else}}
+      <div class="fs-explorer boxed-section">
+        <div class="boxed-section-head">
+          {{fs-breadcrumbs task=task path=path}}
```
Comment (backspace) [inline ui/app/templates/allocations/allocation/task/fs.hbs]: Extraction success! 😀

File: ui/tests/acceptance/task-fs-test.js [hunk 12]
```diff
@@ -256,12 +296,40 @@ module('Acceptance | task fs', function(hooks) {
...
+      firstAttempt.url.split('?')[0],
+      `//${node.httpAddr}/v1/client/fs/readat/${allocation.id}`,
+      'Client is hit first'
+    );
+    assert.equal(firstAttempt.status, 500, 'Client request fails');
+    assert.equal(
+      secondAttempt.url.split('?')[0],
+      `/v1/client/fs/readat/${allocation.id}`,
+      'Server is hit second'
+    );
```
Comment (backspace) [inline ui/tests/acceptance/task-fs-test.js]: nice additions!

File: ui/tests/acceptance/task-fs-test.js [hunk 13]
```diff
@@ -256,12 +296,40 @@ module('Acceptance | task fs', function(hooks) {
   });
 
   test('viewing a file', async function(assert) {
+    const node = server.db.nodes.find(allocation.nodeId);
+
+    server.logging = true;
```
Comment (backspace) [inline ui/tests/acceptance/task-fs-test.js]: Is this meant to be left here?

Comment (DingoEatingFuzz) [reply ui/tests/acceptance/task-fs-test.js]: Hahaha, this is extra funny considering I'm removing it elsewhere in the same PR.

**Comment by DingoEatingFuzz** [comment]
One last re-review? 🤞 

I addressed the character encoding issue and fixed the flapping tests (which were due to illegal characters I added to further test the encoding issue).

File: ui/out [hunk 14]
```diff
@@ -0,0 +1,1317 @@
+Environment: test
```
Comment (backspace) [inline ui/out]: Looks like a stray inclusion of this file?

Comment (DingoEatingFuzz) [reply ui/out]: Lol, I can't win.

**Pull request merged**